multiview/cluster/mv_spectral.py

```diff
@@ -0,0 +1,232 @@
+
+    random_state : int, optional (default=None)
+        Determines random number generation for kmeans.
+
+    info_view : int, optional (default=None)
+        The most informative view. Must be between 0 and n_views-1
+        If given, then the final clustering will be performed on the
+        designated view alone. Otherwise, the algorithm will concatenate
+        across all views and cluster on the result.
+
+    max_iter : int, optional (default=10)
+        The maximum number of iterations to run the clustering
+        algorithm.
+
+    References
+    ----------
+    [1] Abhishek Kumar and Hal Daume. A Co-training Approach for Multiview
+    Spectral Clustering. In International Conference on Machine Learning, 2011
+
+    '''
+
+    def __init__(self, n_clusters, n_views=2, random_state=None,
+                 info_view=None, n_iter=10):
+
+        super().__init__()
+
+        if not (isinstance(n_clusters, int) and n_clusters > 0):
+            msg = 'n_clusters must be a positive integer'
+            raise ValueError(msg)
+
+        if not (isinstance(n_views, int) and n_views > 0):
+            msg = 'n_views must be a positive integer'
+            raise ValueError(msg)
+
+        if random_state is not None:
+            msg = 'random_state must be convertible to 32 bit unsigned integer'
+            try:
+                random_state = int(random_state)
+            except ValueError:
+                raise ValueError(msg)
+            np.random.seed(random_state)
+
+        self.info_view = None
+        if info_view is not None:
+            if (isinstance(info_view, int)
+                    and (info_view >= 0 and info_view < n_views)):
+                self.info_view = info_view
+            else:
+                msg = 'info_view must be an integer between 0 and n_clusters-1'
+                raise ValueError(msg)
+
+        if not (isinstance(n_iter, int) and (n_iter > 0)):
+            msg = 'max_iter must be a positive integer'
+            raise ValueError(msg)
+
+        self.n_clusters = n_clusters
+        self.n_views = n_views
+        self.random_state = random_state
+        self.info_view = info_view
+        self.max_iter = n_iter
+
+    def _gaussian_sim(self, X):
+
+        '''
+        Computes the gaussian similarity kernel for a given matrix.
+        The sigma used is the median pairwise distances.
+
+        Parameters
+        ----------
+        X : array_like, shape(n_samples, n_features)
+            The data matrix from which we will compute the gaussian
+            similarity kernel.
+
+        Returns
+        -------
+        sims : array_like, shape(n_samples, n_samples)
+            The gaussian similarity kernel.
+
+        '''
+
+        distances = cdist(X, X)
+        sq_dists = np.square(distances)
+        norm_dists = sq_dists / (2 * np.median(sq_dists))
+        sims = np.exp(-norm_dists)
+
+        return sims
+
+    def _compute_eigs(self, X):
```

Comment:
1. there must be something in sklearn that already does this entire thing, like i mentioned early in the semester
2. does mv not have a consistent way of doing an embedding? Since you are doing eigendecomp of symmetric PSD matrix, can just use svd, as do many functions in multiview, so I see this as another place where you should either be able to recycle other code from the package or sklearn

Reply:
I replaced the portion involving calculating eigenvectors, but I still need most of that function to compute the normalized graph laplacian.